# B5 — Vectors and Alignment [Optional]

---

## 🔍 Problem-এ কী চাওয়া হয়েছে?

তিনটি vector দেওয়া আছে:

$$\mathbf{a} = [3, -1, 2] \qquad \mathbf{b} = [4, 0, -2] \qquad \mathbf{c} = [-6, 2, -4]$$

তিনটি কাজ করতে হবে:

| Task | কী করতে হবে |
|---|---|
| **(a)** | $\mathbf{a} \cdot \mathbf{b}$ এবং $\mathbf{a} \cdot \mathbf{c}$ হিসাব করা |
| **(b)** | dot product-এর **sign ও magnitude** দেখে $\mathbf{a}$-এর সাথে $\mathbf{b}$ ও $\mathbf{c}$-এর **alignment** বিশ্লেষণ করা |
| **(c)** | $\mathbf{a}$-কে **L2 normalize** করা — result তিন দশমিক পর্যন্ত দেখানো |


---

## 🎯 এই কাজ থেকে আমরা কী অর্জন করতে পারব?

- **Dot product** এর geometric অর্থ বুঝব — শুধু সংখ্যা নয়, দুটো vector কতটা "একই দিকে" আছে সেটা।
- Dot product-এর **sign** (positive/negative/zero) দেখে alignment বিচার করতে পারব।
- **L2 Normalization** কী এবং কেন করা হয় — unit vector-এর ধারণা পাকা হবে।
- Normalized vector-এর **L2 norm সবসময় 1** হয় — এটি verify করব।


---

## 🧠 আমরা যা শিখেছি, সেই আলোকে কীভাবে চিন্তা করতে হবে?

### Dot Product এর Geometric অর্থ:

$$\mathbf{a} \cdot \mathbf{b} = \|\mathbf{a}\| \|\mathbf{b}\| \cos\theta$$

যেখানে $\theta$ হলো দুটো vector-এর মধ্যবর্তী কোণ।

| Dot Product | $\cos\theta$ | কোণ $\theta$ | Alignment |
|---|---|---|---|
| **বড় positive** | > 0, বড় | ছোট (acute) | একই দিকে — **strongly aligned** |
| **ছোট positive** | > 0, ছোট | 0°–90°-এর মধ্যে | কিছুটা একই দিকে |
| **0** | = 0 | ঠিক 90° | **perpendicular** — কোনো alignment নেই |
| **negative** | < 0 | obtuse (90°–180°) | বিপরীত দিকে — **anti-aligned** |

### L2 Normalization:

$$\hat{\mathbf{a}} = \frac{\mathbf{a}}{\|\mathbf{a}\|_2} = \frac{\mathbf{a}}{\sqrt{a_1^2 + a_2^2 + a_3^2}}$$

**কেন normalize করা হয়?**
- Normalized vector-এর **L2 norm = 1** — এটিকে **unit vector** বলে।
- দুটো vector-এর alignment compare করতে চাইলে **magnitude-এর প্রভাব বাদ দিতে** normalize করা হয়।
- ML-এ **cosine similarity** হিসাব করার আগে normalize করা হয়।

### এই problem-এ একটি বিশেষ hint:
$$\mathbf{c} = [-6, 2, -4] = -2 \times [3, -1, 2] = -2\mathbf{a}$$

তাই $\mathbf{c}$ হলো $\mathbf{a}$-এর **exact opposite direction** — এটি dot product-এ স্পষ্ট হবে।


---

## 🛠️ Problem Solve করার Approach

**Step 1:** তিনটি vector define করা।

**Step 2 (Task a):** $\mathbf{a} \cdot \mathbf{b}$ ও $\mathbf{a} \cdot \mathbf{c}$ হাতে হিসাব করা, তারপর code verify।

**Step 3 (Task b):** Sign ও magnitude দেখে alignment বিশ্লেষণ।

**Step 4 (Task c):** $\mathbf{a}$-এর L2 norm বের করে normalize করা।


## Step 1: Vectors Define করা

In [14]:
import numpy as np
import matplotlib.pyplot as plt

a = np.array([3, -1, 2], dtype=float)
b = np.array([4,  0, -2], dtype=float)
c = np.array([-6, 2, -4], dtype=float)

print(f"a = {a}")
print(f"b = {b}")
print(f"c = {c}")
print()
print(f"Notice: c = {c}  =  -2 × {a}  =  -2a")
print(f"Verify: -2 * a = {-2 * a}  →  equals c: {np.array_equal(-2 * a, c)}")


a = [ 3. -1.  2.]
b = [ 4.  0. -2.]
c = [-6.  2. -4.]

Notice: c = [-6.  2. -4.]  =  -2 × [ 3. -1.  2.]  =  -2a
Verify: -2 * a = [-6.  2. -4.]  →  equals c: True


`c = -2 * a` — এটি গুরুত্বপূর্ণ observation।
c হলো a-এর **scalar multiple** — একই line-এ কিন্তু **বিপরীত দিকে** এবং দ্বিগুণ দীর্ঘ।
এই সম্পর্ক dot product-এর result-এ স্পষ্ট হবে।


---

## Task (a): Dot Product — হাতে হিসাব

### $\mathbf{a} \cdot \mathbf{b}$:

$$\mathbf{a} \cdot \mathbf{b} = (3 \times 4) + ((-1) \times 0) + (2 \times (-2))$$
$$= 12 + 0 + (-4) = \mathbf{8}$$

### $\mathbf{a} \cdot \mathbf{c}$:

$$\mathbf{a} \cdot \mathbf{c} = (3 \times (-6)) + ((-1) \times 2) + (2 \times (-4))$$
$$= -18 + (-2) + (-8) = \mathbf{-28}$$


In [15]:
dot_ab_manual = (a[0]*b[0]) + (a[1]*b[1]) + (a[2]*b[2])
dot_ac_manual = (a[0]*c[0]) + (a[1]*c[1]) + (a[2]*c[2])

dot_ab_numpy  = np.dot(a, b)
dot_ac_numpy  = np.dot(a, c)

print(f"a·b  Manual : {dot_ab_manual}")
print(f"a·b  NumPy  : {dot_ab_numpy}")
print(f"Match       : {dot_ab_manual == dot_ab_numpy}")
print()
print(f"a·c  Manual : {dot_ac_manual}")
print(f"a·c  NumPy  : {dot_ac_numpy}")
print(f"Match       : {dot_ac_manual == dot_ac_numpy}")


a·b  Manual : 8.0
a·b  NumPy  : 8.0
Match       : True

a·c  Manual : -28.0
a·c  NumPy  : -28.0
Match       : True


`a[0]*b[0] + a[1]*b[1] + a[2]*b[2]` → element-wise গুণ করে যোগ।
`np.dot(a, b)` → NumPy-র built-in dot product — একই কাজ একটি function-এ।


---

## Task (b): Sign ও Magnitude দেখে Alignment বিশ্লেষণ


In [16]:
norm_a = np.linalg.norm(a)
norm_b = np.linalg.norm(b)
norm_c = np.linalg.norm(c)

print(f"||a|| = {norm_a:.4f}")
print(f"||b|| = {norm_b:.4f}")
print(f"||c|| = {norm_c:.4f}")
print()


# cosine similarity = dot / (||a|| * ||b||)
cos_ab = dot_ab_numpy / (norm_a * norm_b)
cos_ac = dot_ac_numpy / (norm_a * norm_c)

print(f"a·b = {dot_ab_numpy:+.1f}  → positive  → acute angle")
print(f"a·c = {dot_ac_numpy:+.1f} → negative → obtuse angle")
print()


angle_ab = np.degrees(np.arccos(np.clip(cos_ab, -1, 1)))
angle_ac = np.degrees(np.arccos(np.clip(cos_ac, -1, 1)))

print(f"Cosine similarity (a, b) = {cos_ab:.4f}  → angle = {angle_ab:.2f}°")
print(f"Cosine similarity (a, c) = {cos_ac:.4f} → angle = {angle_ac:.2f}°")


||a|| = 3.7417
||b|| = 4.4721
||c|| = 7.4833

a·b = +8.0  → positive  → acute angle
a·c = -28.0 → negative → obtuse angle

Cosine similarity (a, b) = 0.4781  → angle = 61.44°
Cosine similarity (a, c) = -1.0000 → angle = 180.00°


`np.linalg.norm()` কোডটি আসলে ভেক্টরের **দৈর্ঘ্য** বা **ম্যাগনিটিউড** (Magnitude) বের করছে। সহজ কথায়, অরিজিন বা মূলবিন্দু $(0,0)$ থেকে ভেক্টরের শেষ বিন্দু পর্যন্ত সরাসরি দূরত্ব কত, এটি সেটিই হিসাব করে। ডিফল্টভাবে এটি **Euclidean distance** (L2 Norm) এর সূত্র ব্যবহার করে।

`np.arccos()` → cosine থেকে angle বের করে (radians-এ)।
`np.degrees()` → radians → degrees-এ convert।
`np.clip(..., -1, 1)` → floating point error-এর কারণে cosine কখনো ±1 সামান্য ছাড়িয়ে যায়, clip সেটা ঠিক করে।

<div align="center">
  <img src=".\B5-angle of two vectors.jpeg" height="300" width="300"/>
</div>

সহজভাবে বললে, গ্রাফে যদি দুটি তীর চিহ্ন (ভেক্টর) থাকে, তাদের মাঝখানে কত ডিগ্রি ফাঁকা আছে, এটি সেই মানটি দিচ্ছে।

* ১. **`cos_ab`**: এটি দুটি ভেক্টরের কসাইন মান (যেমন $\cos \theta$) নির্দেশ করে।
* ২. **`np.clip(cos_ab, -1, 1)`**: এটি একটি সেফটি চেক। গাণিতিক ত্রুটির কারণে কসাইন মান কখনো $-1$ থেকে $1$ এর সামান্য বাইরে চলে গেলে এটি তাকে সীমার মধ্যে আটকে দেয়।
* ৩. **`np.arccos(...)`**: এটি কসাইন মান থেকে কোণ বা এঙ্গেল বের করে। এটি ফলাফল দেয় **রেডিয়ানে** ($\text{Radians}$)।
* ৪. **`np.degrees(...)`**: রেডিয়ান মানটিকে আমাদের চেনা **ডিগ্রি** (যেমন: $30^\circ$, $45^\circ$, $90^\circ$) ফরম্যাটে রূপান্তর করে।

**সারকথা:** তুমি যদি জানতে চাও ভেক্টর দুটি একে অপরের থেকে কত ডিগ্রি কোণে হেলে আছে, তবে এই কোডটি সেই সরাসরি মানটি ($0^\circ$ থেকে $180^\circ$-এর মধ্যে) তোমাকে দেবে।

**বিশ্লেষণ:**
- $\mathbf{a} \cdot \mathbf{b} = +8$ → **positive** → $\mathbf{b}$ ও $\mathbf{a}$ একই দিকে — **partially aligned**
- $\mathbf{a} \cdot \mathbf{c} = -28$ → **strongly negative** → $\mathbf{c}$ ও $\mathbf{a}$ **বিপরীত দিকে** — angle = 180° (exact opposite)
- Magnitude: $|-28| > |8|$ → $\mathbf{c}$-এর সাথে alignment অনেক বেশি strong (কিন্তু বিপরীত দিকে)


---

## Task (c): L2 Normalize $\mathbf{a}$

$$\|\mathbf{a}\|_2 = \sqrt{3^2 + (-1)^2 + 2^2} = \sqrt{9 + 1 + 4} = \sqrt{14} \approx 3.742$$

$$\hat{\mathbf{a}} = \frac{\mathbf{a}}{\|\mathbf{a}\|_2} = \frac{[3, -1, 2]}{3.742} = \left[\frac{3}{3.742},\ \frac{-1}{3.742},\ \frac{2}{3.742}\right] \approx [0.802,\ -0.267,\ 0.535]$$


In [17]:
l2_norm_a    = np.linalg.norm(a)
a_normalized = a / l2_norm_a

print(f"L2 norm of a      : {l2_norm_a:.4f}  (= sqrt(14) = {np.sqrt(14):.4f})")
print()
print(f"Original a        : {a}")
print(f"Normalized a (â)  : {a_normalized.round(3)}")
print()
# Verify: L2 norm of normalized vector must be 1
norm_of_normalized = np.linalg.norm(a_normalized)
print(f"L2 norm of â      : {norm_of_normalized:.10f}  (expected: 1.0)")
print(f"Is unit vector    : {np.isclose(norm_of_normalized, 1.0)}")


L2 norm of a      : 3.7417  (= sqrt(14) = 3.7417)

Original a        : [ 3. -1.  2.]
Normalized a (â)  : [ 0.802 -0.267  0.535]

L2 norm of â      : 1.0000000000  (expected: 1.0)
Is unit vector    : True


`a / l2_norm_a` → প্রতিটি element-কে L2 norm দিয়ে ভাগ করা (NumPy broadcasting)।
`np.linalg.norm(a_normalized)` → normalized vector-এর norm verify করা।

**গুরুত্বপূর্ণ property:** যেকোনো non-zero vector-এর L2 norm দিয়ে ভাগ করলে result-এর L2 norm সবসময় **ঠিক 1** হয়।
এটিকে **unit vector** বলে — শুধু direction আছে, magnitude নেই।


## Final Summary

In [18]:
print("=" * 55)
print("           COMPLETE RESULTS SUMMARY")
print("=" * 55)
print(f"  a = {a}")
print(f"  b = {b}")
print(f"  c = {c}")
print("-" * 55)
print(f"  a·b  =  {dot_ab_numpy:+.1f}  → positive  → partially aligned")
print(f"  a·c  =  {dot_ac_numpy:+.1f} → negative → anti-aligned (180°)")
print("-" * 55)
print(f"  Angle(a, b)  =  {angle_ab:.2f}°")
print(f"  Angle(a, c)  =  {angle_ac:.2f}°")
print("-" * 55)
print(f"  L2 norm of a =  {l2_norm_a:.4f}")
print(f"  Normalized â =  {a_normalized.round(3)}")
print(f"  ||â||        =  {np.linalg.norm(a_normalized):.3f}  ✅ (unit vector)")
print("=" * 55)


           COMPLETE RESULTS SUMMARY
  a = [ 3. -1.  2.]
  b = [ 4.  0. -2.]
  c = [-6.  2. -4.]
-------------------------------------------------------
  a·b  =  +8.0  → positive  → partially aligned
  a·c  =  -28.0 → negative → anti-aligned (180°)
-------------------------------------------------------
  Angle(a, b)  =  61.44°
  Angle(a, c)  =  180.00°
-------------------------------------------------------
  L2 norm of a =  3.7417
  Normalized â =  [ 0.802 -0.267  0.535]
  ||â||        =  1.000  ✅ (unit vector)


### Key Takeaways:

| | $\mathbf{a}$ vs $\mathbf{b}$ | $\mathbf{a}$ vs $\mathbf{c}$ |
|---|---|---|
| **Dot product** | +8 (positive) | −28 (negative) |
| **Alignment** | Partially aligned | Exactly anti-aligned |
| **Angle** | ~64° | 180° |
| **কারণ** | কিছু element একই দিকে | $\mathbf{c} = -2\mathbf{a}$ — exact opposite |

**L2 Normalization:** $\hat{\mathbf{a}} = [0.802, -0.267, 0.535]$ — magnitude বাদ দিয়ে শুধু direction রাখা হয়েছে।


---